### PR TITLE
iplog rework + omapi integration

### PR DESCRIPTION
**Pull request by dwlfrth**
## Description

Retrieve the mac of an ip from the dhcpd server using omapi
This also include iplog rework
## Impacts

mac2ip and ip2mac lookup 
## NEWS file entries

New Features
++++++++++++
Can optionally query dhcpd server for mac address of the ip address
## Delete branch after merge

YES

### CODE REVIEW

File: lib/pf/iplog.pm [hunk 1]
```diff
@@ -20,15 +20,14 @@ Read the F<pf.conf> configuration file.
 use strict;
 use warnings;
 
-use Net::Netmask;
-use Net::Ping;
 use Date::Parse;
 use Log::Log4perl;
 use Log::Log4perl::Level;
 use IO::Interface::Simple;
-use Net::ARP;
```
Comment (fdurand) [inline lib/pf/iplog.pm:29]: Don't we have to remove them in packetfence.spec and debian/control ?

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 2]
```diff
@@ -11,3 +11,51 @@ CREATE TABLE keyed (
   value LONGBLOB,
   PRIMARY KEY(id)
 ) ENGINE=InnoDB;
+
+--
+-- Table structure for table `iplog_old`
+--
+
+CREATE TABLE iplog_old (
+  mac varchar(255) NOT NULL,
```
Comment (julsemaan) [inline db/upgrade-X.X.X-X.Y.Z.sql]: I don't see the point of using a 255 length for a fixed length mac address

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 3]
```diff
@@ -11,3 +11,51 @@ CREATE TABLE keyed (
...
+  ip varchar(255) NOT NULL,
```
Comment (julsemaan) [inline db/upgrade-X.X.X-X.Y.Z.sql]: Same here, we should lower that value to the maximum char length of an ipv4 address

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 4]
```diff
@@ -11,3 +11,51 @@ CREATE TABLE keyed (
...
+  start_time datetime NOT NULL,
+  end_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP
+) ENGINE=InnoDB;
+
+--
+-- Table structure for table 'iplog'
+--
+
+ALTER TABLE iplog MODIFY mac varchar(255) NOT NULL;
```
Comment (julsemaan) [inline db/upgrade-X.X.X-X.Y.Z.sql]: same here for length

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 5]
```diff
@@ -11,3 +11,51 @@ CREATE TABLE keyed (
...
+ALTER TABLE iplog MODIFY ip varchar(255) NOT NULL;
```
Comment (julsemaan) [inline db/upgrade-X.X.X-X.Y.Z.sql]: same here for length

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 6]
```diff
@@ -11,3 +11,51 @@ CREATE TABLE keyed (
...
+ALTER TABLE iplog DROP INDEX iplog_end_time;
+ALTER TABLE iplog DROP FOREIGN KEY 0_63;
+
+--
+-- Trigger to insert old record from 'iplog' in 'iplog_old' before updating the current one
+--
+
+DROP TRIGGER IF EXISTS iplog_insert_iplog_old_before_update_trigger;
+DELIMITER /
+CREATE TRIGGER iplog_insert_iplog_old_before_update_trigger BEFORE UPDATE ON iplog
```
Comment (julsemaan) [inline db/upgrade-X.X.X-X.Y.Z.sql]: That trigger is nice and readable. Me very like

File: lib/pf/OMAPI.pm [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+
+=head1 DESCRIPTION
+
+pf::OMAPI
+
+=head1 SYNOPSIS
+
+Quick summary of what the module does.
+
+Perhaps a little code snippet.
```
Comment (julsemaan) [inline lib/pf/OMAPI.pm]: Perhaps ? I see it below

File: lib/pf/OMAPI.pm [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+    'state'         => 'N',
+    'atsfp'         => 'N',
+    'starts'        => 'N',
+    'subnet'        => 'N',
+    'hardware-type' => 'N',
+    'result'        => 'N',
+);
+
+
+#The methods used to unpack special values
```
Comment (julsemaan) [inline lib/pf/OMAPI.pm:178]: Shouldn't this be an item in poddoc ?

File: lib/pf/OMAPI.pm [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+#
+our %UNPACK_DATA = (
+    'ip-address' => , \&unpack_ip_address,
+    'hardware-address' =>,\&unpack_hardware_address,
+);
+
+#The methods used to pack special values
```
Comment (julsemaan) [inline lib/pf/OMAPI.pm:185]: same here for poddoc ?

File: lib/pf/iplog.pm [hunk 10]
```diff
@@ -267,91 +269,105 @@ sub iplog_expire {
...
+=cut
+
 sub ip2mac {
-    my ( $ip, $date ) = @_;
-    my $logger = Log::Log4perl::get_logger('pf::iplog');
-    my $mac;
+    my ( $ip ) = @_;
+    my $logger = pf::log::get_logger();
+
     return (0) if ( !valid_ip($ip) );
```
Comment (julsemaan) [inline lib/pf/iplog.pm]: I know you didn't write that piece, but an error or warn log message would be useful when returning here

File: lib/pf/iplog.pm [hunk 11]
```diff
@@ -267,91 +269,105 @@ sub iplog_expire {
...
 }
 
-sub ip2macinarp {
+=head2 ip2macomapi
+
+Look for the MAC address of a given IP address in the DHCP leases using OMAPI
+
+=cut
+
+sub ip2macomapi {
```
Comment (julsemaan) [inline lib/pf/iplog.pm]: this looks a bit unreadable to my eyes. A little underscore wouldn't hurt here

File: lib/pf/iplog.pm [hunk 12]
```diff
@@ -267,91 +269,105 @@ sub iplog_expire {
...
+    return $data->{'obj'}{'hardware-address'} if defined $data;
+}
+
+=head2 ip2macsql
+
+Look for the MAC address of a given IP address using the SQL 'iplog' table
+
+=cut
+
+sub ip2macsql {
```
Comment (julsemaan) [inline lib/pf/iplog.pm]: little underscore here too

File: lib/pf/iplog.pm [hunk 13]
```diff
@@ -267,91 +269,105 @@ sub iplog_expire {
...
+
+sub iplogCache { pf::CHI->new(namespace => 'iplog') }
+
+=head2 mac2ipomapi
+
+Look for the ip in the dhcpd lease entry using omapi
+
+=cut
+
+sub mac2ipomapi {
```
Comment (julsemaan) [inline lib/pf/iplog.pm]: here too ?

File: lib/pf/iplog.pm [hunk 14]
```diff
@@ -374,6 +390,63 @@ sub mac2ip {
...
+sub _lookup_cached_omapi {
+    my ($type, $id) = @_;
+    my $cache = iplogCache();
+    return $cache->compute(
+        $id,
+        {expire_if => \&_expire_lease, expires_in => IPLOG_CACHE_EXPIRE},
+        sub {
+            my $data = _get_lease_from_omapi($type, $id);
+            return unless $data && $data->{op} == 3;
+            #Do not return the lease if it has expires
```
Comment (julsemaan) [inline lib/pf/iplog.pm]: *if it has expired

File: lib/pf/services/manager/dhcpd.pm [hunk 15]
```diff
@@ -87,6 +88,32 @@ EOT
...
+
+=head2 omapi_section
+
+Generate the omapi section if it is defined
+
+=cut
+
+sub omapi_section {
+    my $omapi_section = $Config{omapi};
+    return '"# OMAPI is not enabled on this server' if isdisabled($omapi_section->{ip2mac_lookup});
```
Comment (julsemaan) [inline lib/pf/services/manager/dhcpd.pm]: That is a good idea. +1

File: lib/pf/OMAPI.pm (hunk 8 above)
Comment (jrouzierinverse) [inline lib/pf/OMAPI.pm:178]: This is internal to the module.
%UNPACK_DATA will not be used outside of the module.

File: lib/pf/OMAPI.pm (hunk 9 above)
Comment (jrouzierinverse) [inline lib/pf/OMAPI.pm:185]: This is internal to the module.
A comment is enough

File: lib/pf/iplog.pm (hunk 1 above)
Comment (dwlfrth) [inline lib/pf/iplog.pm:29]: I'd have to check if they are used elsewhere in the codebase...

File: db/upgrade-X.X.X-X.Y.Z.sql (hunk 3 above)
Comment (dwlfrth) [inline db/upgrade-X.X.X-X.Y.Z.sql]: I embrace ipv6 !

File: db/upgrade-X.X.X-X.Y.Z.sql (hunk 5 above)
Comment (dwlfrth) [inline db/upgrade-X.X.X-X.Y.Z.sql]: I embrace ipv6!

File: db/pf-schema-X.Y.Z.sql [hunk 16]
```diff
@@ -208,24 +208,49 @@ CREATE TABLE violation (
...
+--
+-- Table structure for table `iplog_archive`
+--
+
+CREATE TABLE iplog_archive (
+  mac varchar(17) NOT NULL,
+  ip varchar(45) NOT NULL,
   start_time datetime NOT NULL,
-  end_time datetime default "0000-00-00 00:00:00"
+  end_time datetime NOT NULL
```
Comment (louismunro) [inline db/pf-schema-X.Y.Z.sql]: No indexes?

File: db/pf-schema-X.Y.Z.sql [hunk 17]
```diff
@@ -208,24 +208,49 @@ CREATE TABLE violation (
...
 -- Table structure for table `iplog_history`
 --
 
 CREATE TABLE iplog_history (
   mac varchar(17) NOT NULL,
-  ip varchar(15) NOT NULL,
+  ip varchar(45) NOT NULL,
+  start_time datetime NOT NULL,
+  end_time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP
+) ENGINE=InnoDB;
```
Comment (louismunro) [inline db/pf-schema-X.Y.Z.sql:243]: No indexes?

File: db/upgrade-X.X.X-X.Y.Z.sql [hunk 18]
```diff
@@ -13,3 +13,63 @@ CREATE TABLE keyed (
...
+-- Table structure for table 'iplog'
+--
+
+ALTER TABLE iplog MODIFY mac varchar(17) NOT NULL;
+ALTER TABLE iplog MODIFY ip varchar(45) NOT NULL;
+ALTER TABLE iplog ADD PRIMARY KEY(ip);
+ALTER TABLE iplog DROP INDEX ip_view_open;
+ALTER TABLE iplog DROP INDEX mac_view_open;
+ALTER TABLE iplog DROP INDEX iplog_end_time;
+ALTER TABLE iplog DROP FOREIGN KEY 0_63;
```
Comment (louismunro) [inline db/upgrade-X.X.X-X.Y.Z.sql:50]: Can these be turned into a single ALTER TABLE?
That might be faster on larger tables.

File: lib/pf/OMAPI.pm [hunk 19]
```diff
@@ -0,0 +1,543 @@
...
+has key_base64 => (is => 'rw');
+
+#The different message types
+
+our $OPEN    = 1;
+our $REFRESH = 2;
+our $UPDATE  = 3;
+our $NOTIFY  = 4;
+our $ERROR   = 5;
+our $DELETE  = 6;
```
Comment (louismunro) [inline lib/pf/OMAPI.pm:156]: Readonly?

File: lib/pf/iplog.pm [hunk 20]
```diff
@@ -193,55 +407,141 @@ sub iplog_view_open_mac {
     return ($ref);
 }
 
-sub iplog_view_all_open_mac {
-    my ($mac) = @_;
-    return db_data(IPLOG, $iplog_statements, 'iplog_view_open_mac_sql', $mac);
+sub list_open {
+    my ( $search_by ) = @_;
+    my $logger = pf::log::get_logger;
```
Comment (louismunro) [inline lib/pf/iplog.pm:488]: Please remove if unused.

File: lib/pf/iplog.pm [hunk 21]
```diff
@@ -193,55 +407,141 @@ sub iplog_view_open_mac {
...
+    return _list_open_by_mac($search_by) if ( defined($search_by) && valid_mac($search_by) );
+
+    return _list_open_by_ip($search_by) if ( defined($search_by) && valid_ip($search_by) );
+
+    return db_data(IPLOG, $iplog_statements, 'iplog_list_open_sql') if ( !defined($search_by) );
+}
+
+sub _list_open_by_ip {
+    my ( $ip ) = @_;
+    my $logger = pf::log::get_logger;
```
Comment (louismunro) [inline lib/pf/iplog.pm:511]: Please remove if unused.

File: lib/pf/iplog.pm [hunk 22]
```diff
@@ -193,55 +407,141 @@ sub iplog_view_open_mac {
...
+
+    return db_data(IPLOG, $iplog_statements, 'iplog_list_open_by_ip_sql', $ip);
+}
+
+sub _list_open_by_mac {
+    my ( $mac ) = @_;
+    my $logger = pf::log::get_logger;
```
Comment (louismunro) [inline lib/pf/iplog.pm:528]: Please remove if unused.

**Comment by louismunro** [comment]
LGTM.

**Pull request merged**